lib/services/systemd.c

```diff
@@ -820,6 +820,80 @@ process_unit_method_reply(DBusMessage *reply, svc_action_t *op)
     }
 }
 
+static DBusHandlerResult
+filter_job_removed_signal(DBusConnection *connection, DBusMessage *message,
+                          void *user_data)
+{
+    svc_action_t *action = user_data;
+    uint32_t job_id = 0;
+    const char *bus_path = NULL;
+    const char *unit_name = NULL;
+    const char *result = NULL;
+    DBusError error;
+
+    CRM_CHECK((connection != NULL) && (message != NULL),
+              return DBUS_HANDLER_RESULT_NOT_YET_HANDLED);
+
+    if ((action == NULL)
+        || !dbus_message_is_signal(message, BUS_NAME_MANAGER, "JobRemoved")) {
+        return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
+    }
+
+    dbus_error_init(&error);
+    if (!dbus_message_get_args(message, &error,
+                               DBUS_TYPE_UINT32, &job_id,
+                               DBUS_TYPE_OBJECT_PATH, &bus_path,
+                               DBUS_TYPE_STRING, &unit_name,
+                               DBUS_TYPE_STRING, &result,
+                               DBUS_TYPE_INVALID)) {
+        crm_err("Could not interpret systemd DBus signal: %s " QB_XS " (%s)",
+                error.message, error.name);
+        dbus_error_free(&error);
+        return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
+    }
+
+    if (!pcmk__str_eq(bus_path, action->opaque->job_path, pcmk__str_none)) {
+        // This filter is not for this job
+        return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
+    }
+
+    crm_trace("Setting %s result for %s (JobRemoved id=%" PRIu32 ", result=%s",
+              pcmk__s(action->action, "(unknown)"), unit_name, job_id, result);
+
+    if (pcmk__str_eq(result, "done", pcmk__str_none)) {
+        services__set_result(action, PCMK_OCF_OK, PCMK_EXEC_DONE, NULL);
```

Comment:
This is new. Previously, I don't see where we were ever setting the `svc_action_t` result for systemd actions that completed. We *did* set the `cmd->result` for the `lrmd_cmd_t` in the executor.

I hope it's safe to set this here. It's yet another reason we should try to test thoroughly.

My guess is that the `svc_action_t` result simply went unused, and all we cared about was the `cmd->result` in the executor. But I'm not confident in that.

Anyway, I'm calling `services__set_result()` on the action for convenience and for logical consistency. Then I'm copying that result in the executor's `action_complete()` callback. It would be trivial to use `pcmk__set_result(&(cmd->result), ...)` in `action_complete()` instead, if this is a problem.

Reply:
> Previously, I don't see where we were ever setting the `svc_action_t` result for systemd actions that completed.

Not sure how I missed it, but this was being done in `process_unit_method_complete()`.